gnovm/stdlibs/std/native.go

```diff
@@ -32,6 +32,37 @@ func GetHeight(m *gno.Machine) int64 {
 	return m.Context.(ExecContext).Height
 }
 
+// getPrevFunctionNameFromTarget returns the last called function name (identifier) from the call stack.
+func getPrevFunctionNameFromTarget(m *gno.Machine, targetFunc string) string {
+	targetIndex := findTargetFuncIndex(m, targetFunc)
+	if targetIndex == -1 {
+		return ""
+	}
+	return findPreviousFuncName(m, targetIndex)
+}
+
+// findTargetFuncIndex finds and returns the index of the target function in the call stack.
+func findTargetFuncIndex(m *gno.Machine, targetFunc string) int {
+	for i := len(m.Frames) - 1; i >= 0; i-- {
+		currFunc := m.Frames[i].Func
+		if currFunc != nil && currFunc.Name == gno.Name(targetFunc) {
+			return i
+		}
+	}
+	return -1
+}
+
+// findPreviousFuncName returns the function name before the given index in the call stack.
+func findPreviousFuncName(m *gno.Machine, targetIndex int) string {
+	for i := targetIndex - 1; i >= 0; i-- {
+		currFunc := m.Frames[i].Func
+		if currFunc != nil {
+			return string(currFunc.Name)
+		}
+	}
+	return ""
+}
```

Comment:
Should these panic in case of an error, and not return `""`?

cc @moul

Reply:
The `findPreviousFuncName` function can use panic, but the `getPrevFunctionNameFromTarget` function should return an empty string.

Especially in the case of test functions like:

```go
func Test_foo(t *testing.T) {
 X_emit(...)
}
```

Since the name of the function calling X_emit is not recognized, it should be handled as an empty string to avoid issues in testing.